### PR TITLE
[WIP] Operation cancellation support

### PR DESCRIPTION
**Pull request by carlosmn**
This is a possible implementation of #3334, but closer to the .NET design with an object that you pass into the operations which cannot be used to request cancellation.

This needs some naming bikeshedding and possibly more exact docs and a PoC implementation in e.g. the network or checkout operations.

It also needs a well-crafted history and commit messages.

### CODE REVIEW

File: include/git2/cancellable.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ */
+typedef int (*git_cancellable_cb)(git_cancellable *cancellable, void *payload);
+
+/**
+ * Create a cancellable source
+ *
+ * Create a new cancellable source. You can pass its token to functions which
+ * accept them and cancel them.
+ *
+ * @param out ponter in which to store the cancellable source
```
Comment (Therzok) [inline include/git2/cancellable.h]: Nit: typo pointer.

**Comment by Therzok** [comment]
Looks good to me!

File: src/cancellable.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	/* Should we have a memory barrier here? a full lock? */
+	return c->cancelled;
+}
+
+int git_cancellable_register(git_cancellable *c, git_cancellable_cb cb, void *payload)
+{
+	registration *reg = git_array_alloc(c->registrations);
+	GITERR_CHECK_ALLOC(reg);
+
+	reg->cb = cb;
```
Comment (Therzok) [inline src/cancellable.c]: This should be executed immediately if it's already been canceled.

**Comment by Therzok** [comment]
Would this proposal include timeout support to `git_cancellable`?

File: src/cancellable.c (hunk 2 above)
Comment (carlosmn) [inline src/cancellable.c]: Ah yes, we can just skip it in that case.

**Comment by carlosmn** [comment]
Adding built-in support for timeouts is something I've thought about and can be added later if we want to. But the primary/first user of this would be libgit2sharp, and we can piggy-back off of its `CancellationToken` so a timeout cancellation token there would make us cancel.

So we would only need it for C itself or some other language which doesn't have an equivalent.

**Comment by carlosmn** [comment]
I've updated the name and removed the distinction between token and source.

**Comment by ethomson** [comment]
Closing as stale.

**Pull request closed without merging**